### DISCUSSION TITLE
DSL: order insensitivity

### DISCUSSION THREAD

**Discussion by vedeja** [structurizr/java#260] [Ideas]
I’m no language designer expert so bear with me if I use the terminology wrong. The order in which the objects are defined in the model is important. It is at is the model is parsed procedurally instead of compiled. This has many consequences, such as many relationships having to be defined last in the model which makes them hard to overview and maintain. I would really like to be able to define all relationships nested (as children) on their source element.

**Comment by simonbrowndotje** [comment]
Correct, this is by design, with the DSL being imperative rather than declarative (https://docs.structurizr.com/dsl/basics). You've probably seen that LikeC4 supports this feature (https://likec4.dev/docs/dsl/references/), but it's not something that will be added to Structurizr. The DSL parser is open source though, so you're welcome to fork it and add forward referencing via something like a double-pass parsing loop. I know of a couple of teams who have done this, but the results are subpar, with features such as implied relationships, scripts, and plugins being particularly negatively affected by the change.

**Reply by vedeja** [reply]
In your opinion, what is the primary benefit of an imperative DSL?

**Reply by simonbrowndotje** [reply]
The DSL is a wrapper over the top of the original Structurizr for Java library, which also takes an imperative approach to creating an architecture model -> https://github.com/structurizr/examples/blob/main/java/src/main/java/com/structurizr/example/GettingStarted.java

The primary benefit is you get all of the features/power of the Java library.
